File: artifacts/api-server/src/routes/returns.ts
import { Router } from "express";
import { eq, and, sql } from "drizzle-orm";
import { db, returnsTable, productsTable, warehousesTable, inventoryBalanceTable, stockMovementsTable, inventoryLedgerTable } from "@workspace/db";
import { ListReturnsQueryParams, CreateReturnBody, UpdateReturnQcParams, UpdateReturnQcBody } from "@workspace/api-zod";
import { randomUUID } from "crypto";

const router = Router();

const withJoins = async (whereClause: Parameters<typeof db.select>[0] extends never ? never : any, page: number, limit: number) => {
  return db.select({
    id: returnsTable.id,
    returnNumber: returnsTable.returnNumber,
    orderId: returnsTable.orderId,
    productId: returnsTable.productId,
    warehouseId: returnsTable.warehouseId,
    qty: returnsTable.qty,
    returnReason: returnsTable.returnReason,
    qcStatus: returnsTable.qcStatus,
    createdAt: returnsTable.createdAt,
    product: {
      id: productsTable.id,
      skuCode: productsTable.skuCode,
      productName: productsTable.productName,
      category: productsTable.category,
      brand: productsTable.brand,
      unit: productsTable.unit,
      minimumStock: productsTable.minimumStock,
      reorderStock: productsTable.reorderStock,
      isActive: productsTable.isActive,
      createdAt: productsTable.createdAt,
      updatedAt: productsTable.updatedAt,
    },
    warehouse: {
      id: warehousesTable.id,
      warehouseName: warehousesTable.warehouseName,
      location: warehousesTable.location,
      warehouseType: warehousesTable.warehouseType,
      createdAt: warehousesTable.createdAt,
    },
  })
  .from(returnsTable)
  .innerJoin(productsTable, eq(returnsTable.productId, productsTable.id))
  .innerJoin(warehousesTable, eq(returnsTable.warehouseId, warehousesTable.id))
  .where(whereClause)
  .orderBy(sql`${returnsTable.createdAt} desc`)
  .limit(limit)
  .offset((page - 1) * limit);
};

router.get("/returns", async (req, res): Promise<void> => {
  const query = ListReturnsQueryParams.safeParse(req.query);
  if (!query.success) {
    res.status(400).json({ error: query.error.message });
    return;
  }
  const { qcStatus, page = 1, limit = 50 } = query.data;
  const whereClause = qcStatus ? eq(returnsTable.qcStatus, qcStatus) : undefined;

  const [data, countResult] = await Promise.all([
    withJoins(whereClause, page, limit),
    db.select({ count: sql<number>`count(*)::int` }).from(returnsTable).where(whereClause),
  ]);

  res.json({ data, total: countResult[0].count, page, limit });
});

router.post("/returns", async (req, res): Promise<void> => {
  const parsed = CreateReturnBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }
  const { productId, warehouseId, qty, returnReason, orderId } = parsed.data;
  const returnNumber = `RET-${Date.now()}`;

  const [ret] = await db.insert(returnsTable).values({
    returnNumber,
    orderId: orderId ?? null,
    productId,
    warehouseId,
    qty,
    returnReason,
    qcStatus: "PENDING",
  }).returning();

  // Move items to RETURN category in inventory
  const movementId = randomUUID();
  let [balance] = await db.select().from(inventoryBalanceTable).where(
    and(eq(inventoryBalanceTable.productId, productId), eq(inventoryBalanceTable.warehouseId, warehouseId))
  );
  if (!balance) {
    const [created] = await db.insert(inventoryBalanceTable).values({ productId, warehouseId }).returning();
    balance = created;
  }
  const openingQty = balance.availableQty;
  const closingQty = openingQty; // returns go to returnedQty, not availableQty

  await db.insert(stockMovementsTable).values({
    movementId,
    productId,
    warehouseId,
    movementType: "RETURN",
    qty,
    referenceType: "return",
    referenceId: returnNumber,
  });

  await db.update(inventoryBalanceTable).set({ returnedQty: balance.returnedQty + qty }).where(
    and(eq(inventoryBalanceTable.productId, productId), eq(inventoryBalanceTable.warehouseId, warehouseId))
  );

  await db.insert(inventoryLedgerTable).values({
    productId,
    warehouseId,
    movementId,
    openingQty,
    changeQty: 0,
    closingQty,
  });

  const [result] = await withJoins(eq(returnsTable.id, ret.id), 1, 1);
  res.status(201).json(result);
});

router.patch("/returns/:id/qc", async (req, res): Promise<void> => {
  const params = UpdateReturnQcParams.safeParse(req.params);
  if (!params.success) {
    res.status(400).json({ error: params.error.message });
    return;
  }
  const parsed = UpdateReturnQcBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }

  const [ret] = await db.select().from(returnsTable).where(eq(returnsTable.id, params.data.id));
  if (!ret) {
    res.status(404).json({ error: "Return not found" });
    return;
  }

  const { qcStatus } = parsed.data;
  const [updated] = await db.update(returnsTable).set({ qcStatus }).where(eq(returnsTable.id, ret.id)).returning();

  // Create appropriate movement based on QC outcome
  const movementId = randomUUID();
  const movementType = qcStatus === "GOOD" ? "INWARD" : qcStatus === "DAMAGED" ? "DAMAGE" : "EXPIRED";

  let [balance] = await db.select().from(inventoryBalanceTable).where(
    and(eq(inventoryBalanceTable.productId, ret.productId), eq(inventoryBalanceTable.warehouseId, ret.warehouseId))
  );
  if (!balance) {
    const [created] = await db.insert(inventoryBalanceTable).values({ productId: ret.productId, warehouseId: ret.warehouseId }).returning();
    balance = created;
  }

  const openingQty = balance.availableQty;
  const changeQty = qcStatus === "GOOD" ? ret.qty : 0;
  const closingQty = openingQty + changeQty;

  await db.insert(stockMovementsTable).values({
    movementId,
    productId: ret.productId,
    warehouseId: ret.warehouseId,
    movementType,
    qty: ret.qty,
    referenceType: "return",
    referenceId: ret.returnNumber,
  });

  const balanceUpdate: Record<string, number> = {};
  if (qcStatus === "GOOD") balanceUpdate.availableQty = closingQty;
  if (qcStatus === "DAMAGED") balanceUpdate.damagedQty = balance.damagedQty + ret.qty;
  if (qcStatus === "EXPIRED") balanceUpdate.expiredQty = balance.expiredQty + ret.qty;
  balanceUpdate.returnedQty = Math.max(0, balance.returnedQty - ret.qty);

  await db.update(inventoryBalanceTable).set(balanceUpdate).where(
    and(eq(inventoryBalanceTable.productId, ret.productId), eq(inventoryBalanceTable.warehouseId, ret.warehouseId))
  );

  await db.insert(inventoryLedgerTable).values({
    productId: ret.productId,
    warehouseId: ret.warehouseId,
    movementId,
    openingQty,
    changeQty,
    closingQty,
  });

  const [result] = await withJoins(eq(returnsTable.id, ret.id), 1, 1);
  res.json(result);
});

export default router;

File: artifacts/api-server/routers/returns.py
import uuid
import time
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import Optional
from database import get_db
from models import Return, InventoryBalance, StockMovement, InventoryLedger
from schemas import ReturnOut, ReturnCreate, ReturnQcUpdate

router = APIRouter(prefix="/api/returns", tags=["returns"])


def _ensure_balance(db, product_id, warehouse_id):
    balance = db.query(InventoryBalance).filter(
        InventoryBalance.product_id == product_id,
        InventoryBalance.warehouse_id == warehouse_id,
    ).first()
    if not balance:
        balance = InventoryBalance(product_id=product_id, warehouse_id=warehouse_id)
        db.add(balance)
        db.flush()
    return balance


@router.get("")
def list_returns(
    qc_status: Optional[str] = None,
    page: int = Query(1, ge=1),
    limit: int = Query(50, ge=1, le=200),
    db: Session = Depends(get_db),
):
    q = db.query(Return)
    if qc_status:
        q = q.filter(Return.qc_status == qc_status)

    total = q.count()
    items = q.order_by(Return.created_at.desc()).offset((page - 1) * limit).limit(limit).all()
    return {"data": [ReturnOut.model_validate(r) for r in items], "total": total, "page": page, "limit": limit}


@router.post("", status_code=201)
def create_return(body: ReturnCreate, db: Session = Depends(get_db)):
    return_number = f"RET-{int(time.time() * 1000)}"
    ret = Return(
        return_number=return_number,
        product_id=body.product_id,
        warehouse_id=body.warehouse_id,
        qty=body.qty,
        return_reason=body.return_reason,
        order_id=body.order_id,
        qc_status="PENDING",
    )
    db.add(ret)
    db.flush()

    balance = _ensure_balance(db, body.product_id, body.warehouse_id)
    movement_id = str(uuid.uuid4())
    db.add(StockMovement(
        movement_id=movement_id,
        product_id=body.product_id,
        warehouse_id=body.warehouse_id,
        movement_type="RETURN",
        qty=body.qty,
        reference_type="return",
        reference_id=return_number,
    ))
    balance.returned_qty += body.qty
    db.add(InventoryLedger(
        product_id=body.product_id,
        warehouse_id=body.warehouse_id,
        movement_id=movement_id,
        opening_qty=balance.available_qty,
        change_qty=0,
        closing_qty=balance.available_qty,
    ))
    db.commit()
    db.refresh(ret)
    return ReturnOut.model_validate(ret)


@router.patch("/{id}/qc")
def update_return_qc(id: int, body: ReturnQcUpdate, db: Session = Depends(get_db)):
    ret = db.query(Return).filter(Return.id == id).first()
    if not ret:
        raise HTTPException(status_code=404, detail="Return not found")

    ret.qc_status = body.qc_status
    balance = _ensure_balance(db, ret.product_id, ret.warehouse_id)

    movement_type = "INWARD" if body.qc_status == "GOOD" else ("DAMAGE" if body.qc_status == "DAMAGED" else "EXPIRED")
    movement_id = str(uuid.uuid4())
    change_qty = ret.qty if body.qc_status == "GOOD" else 0

    db.add(StockMovement(
        movement_id=movement_id,
        product_id=ret.product_id,
        warehouse_id=ret.warehouse_id,
        movement_type=movement_type,
        qty=ret.qty,
        reference_type="return",
        reference_id=ret.return_number,
    ))

    opening_qty = balance.available_qty
    closing_qty = opening_qty + change_qty

    if body.qc_status == "GOOD":
        balance.available_qty = closing_qty
    elif body.qc_status == "DAMAGED":
        balance.damaged_qty += ret.qty
    else:
        balance.expired_qty += ret.qty
    balance.returned_qty = max(0, balance.returned_qty - ret.qty)

    db.add(InventoryLedger(
        product_id=ret.product_id,
        warehouse_id=ret.warehouse_id,
        movement_id=movement_id,
        opening_qty=opening_qty,
        change_qty=change_qty,
        closing_qty=closing_qty,
    ))
    db.commit()
    db.refresh(ret)
    return ReturnOut.model_validate(ret)
